### PR TITLE
Add OSRAM On/Off plug support

### PR DESCRIPTION
**Pull request by herbrechtsmeier**

### CODE REVIEW

**Comment by sieren** [comment]
Probably better to use the `development` branch and add the type to https://github.com/enwi/hueplusplus/blob/development/src/HueDeviceTypes.cpp

**Comment by herbrechtsmeier** [comment]
I have switch the branch but this confuses GitHub.

**Comment by enwi** [comment]
> I have switch the branch but this confuses GitHub.

You need to rebase on to the development branch, because we forced pushed on it and you are diverged.
Try the following:
```
git checkout development
git pull
git checkout feature/add-osram-plug-support
git rebase development
git push --force
```

**Comment by enwi** [comment]
Here I did it for you really quick

Just as a reference for what I did:

```
git checkout development
git remote add upstream git@github.com:enwi/hueplusplus.git
git fetch upstream
git reset --hard upstream/development
git checkout feature/add-osram-plug-support
git rebase development
git push --force
```

Also @herbrechtsmeier what you need to do on you local clone is to call 
```
git pull --rebase
```

**Comment by herbrechtsmeier** [comment]
Why the [HueDeviceTypes](https://github.com/enwi/hueplusplus/blob/development/src/HueDeviceTypes.cpp) doesn't return an `HueLightType` enum to require the type only?

Why the class uses a `std::set` instead of a `std::unordered_map` or the solution from [nlohmann/json](https://github.com/nlohmann/json/blob/f2b43a36b2b7cba3e4e9a1d056c9b83687252407/include/nlohmann/detail/macro_scope.hpp#L81)?

**Comment by herbrechtsmeier** [comment]
My problem was a missing remote `origin` for the `git fetch` command.

**Comment by enwi** [comment]
> My problem was a missing remote `origin` for the `git fetch` command.

Yes, I was following [these steps](https://gist.github.com/glennblock/1974465) and noticed that you needed to add the remote a little late.

**Comment by enwi** [comment]
> Why the [HueDeviceTypes](https://github.com/enwi/hueplusplus/blob/development/src/HueDeviceTypes.cpp) doesn't return an `HueLightType` enum to require the type only?
> 
> Why the class uses a `std::set` instead of a `std::unordered_map` or the solution from [nlohmann/json](https://github.com/nlohmann/json/blob/f2b43a36b2b7cba3e4e9a1d056c9b83687252407/include/nlohmann/detail/macro_scope.hpp#L81)?

Also I would say this should be part of a new issue. But you are right, it might be more elegant to have an enum with all light types

File: src/HueDeviceTypes.cpp [hunk 1]
```diff
@@ -66,6 +66,13 @@ const std::set<std::string> getTemperatureLightTypes()
             "LTC003", "LTC004", "LTC011", "LTC012", "LTD001", "LTD002", "LFF001", "LTT001", "LDT001"};
     return c_TEMPERATURELIGHT_TYPES;
 }
+
+const std::set<std::string> getPlugTypes()
```
Comment (Jojo-1000) [inline src/HueDeviceTypes.cpp:70]: Maybe call this getNonDimmableTypes(), so other lights that are not plugs can be put in here too.

Comment (herbrechtsmeier) [reply src/HueDeviceTypes.cpp:70]: Should I add the new function to the end below `getTemperatureLightTypes` or between `getNoColorTypes` and `getTemperatureLightTypes` ?

**Comment by Jojo-1000** [comment]
> Why the class uses a `std::set` instead of a `std::unordered_map` or the solution from [nlohmann/json](https://github.com/nlohmann/json/blob/f2b43a36b2b7cba3e4e9a1d056c9b83687252407/include/nlohmann/detail/macro_scope.hpp#L81)?

A map would require storing the type _and_ the color type for each type. That would duplicate a lot of data, because many types have the same value.
My idea for MakeHueLight is to store all strategies in the factory, so the Hue bridge gets simpler.

File: src/HueDeviceTypes.cpp (hunk 1 above)
Comment (Jojo-1000) [reply src/HueDeviceTypes.cpp:70]: If you want to sort them, you could do gamut a,b,c, then temperature, then noColor (should probably be renamed to dimmable) and finally non dimmable.

**Comment by herbrechtsmeier** [comment]
Indeed a map would duplicate data and a pair together with a find_if and switch would be possible but probable wouldn't save much.
```
static const std::pair<int, std::set<std::string>> m[] = {
    {1, getGamutBTypes()},
    {2, getGamutCTypes()},
    ...
};

auto it = std::find_if(std::begin(m), std::end(m),
                        [&type](const std::pair<int, std::set<std::string>>& p) -> bool
{
    return p.second.count(type);
});
int e = ((it != std::end(m)) ? it : std::begin(m))->first;

swtich(e)
case 1:
....
```

**Comment by sieren** [comment]
Yeah, this is what I had in mind too when i tried to consolidate this code but ultimately abandoned that idea for the same reason.

**Pull request merged**